libyul/backends/evm/StackHelpers.h

```diff
@@ -0,0 +1,363 @@
+	// Returns true, iff the current slot at sourceOffset in source layout is a suitable slot at targetOffset.
+	{ ops.isCompatible(sourceOffset, targetOffset) } -> std::convertible_to<bool>;
+	// Returns true, iff the slots at the two given source offsets are identical.
+	{ ops.sourceIsSame(sourceOffset, sourceOffset) } -> std::convertible_to<bool>;
+	// Returns a positive integer n, if the slot at the given source offset needs n more copies.
+	// Returns a negative integer -n, if the slot at the given source offsets occurs n times too many.
+	// Returns zero if the amount of occurrences, in the current source layout, of the slot at the given source offset matches the desired amount of occurrences in the target.
+	{ ops.sourceMultiplicity(sourceOffset) } -> std::convertible_to<int>;
+	// Returns a positive integer n, if the slot at the given target offset needs n more copies.
+	// Returns a negative integer -n, if the slot at the given target offsets occurs n times too many.
+	// Returns zero if the amount of occurrences, in the current source layout, of the slot at the given target offset matches the desired amount of occurrences in the target.
+	{ ops.targetMultiplicity(targetOffset) } -> std::convertible_to<int>;
+	// Returns true, iff any slot is compatible with the given target offset.
+	{ ops.targetIsArbitrary(targetOffset) } -> std::convertible_to<bool>;
+	// Returns the number of slots in the source layout.
+	{ ops.sourceSize() } -> std::convertible_to<size_t>;
+	// Returns the number of slots in the target layout.
+	{ ops.targetSize() } -> std::convertible_to<size_t>;
+	// Swaps the top most slot in the source with the slot at depth.
+	{ ops.swap(depth) };
+	// Pops the top most slot in the source.
+	{ ops.pop() };
+	// Dups or pushes the slot that is supposed to end up at the given target offset.
+	{ ops.pushOrDupTarget(targetOffset) };
+};
+*/
+/// Helper class that can perform shuffling of a source stack layout to a target stack layout via
+/// abstracted shuffle operations.
+template</*ShuffleOperationConcept*/ typename ShuffleOperations>
+class Shuffler
+{
+public:
+	/// Executes the stack shuffling operations. Instantiates an instance of ShuffleOperations
+	/// in each iteration. Each iteration performs exactly one operation that modifies the stack.
+	/// After shuffle, source and target have the same size and all slots in the source layout are
+	/// compatible to the slots at the same target offset.
+	template<typename... Args>
+	static void shuffle(Args&&... args)
+	{
+		bool needsMoreShuffling = true;
+		// The shuffling algorithm should always terminate in polynomial time, but we provide a limit
+		// in case it does not terminate due to a bug.
+		size_t iterationCount = 0;
+		while (iterationCount < 1000 && (needsMoreShuffling = shuffleStep(std::forward<Args>(args)...)))
+			++iterationCount;
+		yulAssert(!needsMoreShuffling, "Could not create stack layout after 1000 iterations.");
```

Comment:
Is this really a hard assert or just a "resource exhaustion"?

Reply:
It's rather meant as a safeguard for cases, in which due to a bug the shuffling would oscillate. Not that I expect it to happen at this point, but *if* it happens better to abort at some point than to keep spinning... realistically I'd expect any stack to have been shuffled after 1000 operations (each iteration produces an opcode).